modules/phase_field/src/userobjects/EBSDReader.C

```diff
@@ -186,19 +199,132 @@ EBSDReader::readFile()
+     * no. 4 (2007): 1193-1197.
+     * A 4 by N matrix (Q) is constructed, where N is the number of quaternions.
+     * A weight matrix (W) is created. The eigen vector corresponding to the
+     * maximum eigen value of Q*W*Q' is the weighted average quaternion
+     **/
+
+    // creating a quaternion matrix
+    Eigen::MatrixXd quat_mat(4, quat[i].size());
+    // creating the weight matrix
+    Eigen::MatrixXd weight = Eigen::MatrixXd::Identity(quat[i].size(), quat[i].size());
+
+    unsigned int bin_size;
+    bool data_quality_ok = false;
+    Real total_weight = 0.0;
+
+    for (unsigned int j = 0; j < quat[i].size(); j++)
+    {
+      bin_size = feature_weights[i][quat_to_bin[i][j]];
+      w = quat[i][j].w();
+      x = quat[i][j].x();
+      y = quat[i][j].y();
+      z = quat[i][j].z();
+
+      // instantiating columns of the matrix
+      quat_mat.col(j) << w, x, y, z;
+      // assigning weights to each quaternion
+      weight(j, j) = std::pow(bin_size, _L_norm);
+      total_weight += weight(j, j);
+      /**
+       * If no bin exists which has atleast 50% of total quaternions in a grain
+       * then the EBSD data may not be reliable
+       * Note: The limit 50% is arbitrary
+       */
+      if (bin_size / quat[i].size() > 0.5)
+        data_quality_ok = true;
+    }
+
+    weight = weight / total_weight;
+
+    // throws a warning if EBSD data is not reliable
+    if (!data_quality_ok)
+      _console << COLOR_YELLOW << "EBSD data may not be reliable"
+               << "\n"
+               << COLOR_DEFAULT;
+
+    // compute eigen values and eigen vectors
+    Eigen::EigenSolver<Eigen::MatrixXd> EigenSolver(quat_mat * weight * quat_mat.transpose());
+    Eigen::VectorXd eigen_values = EigenSolver.eigenvalues().real();
+    Eigen::MatrixXd eigen_vectors = EigenSolver.eigenvectors().real();
+
+    // creating a tuple of eigen values and eigen vectors
+    for (unsigned int i = 0; i < eigen_values.size(); i++)
+    {
+      std::tuple<double, Eigen::VectorXd> vec_and_val(eigen_values[i], eigen_vectors.col(i));
+      eigen_vectors_and_values.push_back(vec_and_val);
+    }
+    // Sorting the eigen values and vectors in ascending order of eigen values
+    auto j = std::max_element(eigen_vectors_and_values.begin(),
+                              eigen_vectors_and_values.end(),
+                              [&](const std::tuple<double, Eigen::VectorXd> & a,
+                                  const std::tuple<double, Eigen::VectorXd> & b) -> bool {
+                                return std::get<0>(a) < std::get<0>(b);
+                              });
+    // Selecting eigen vector corresponding to max eigen value to compute average euler angle
+    Eigen::VectorXd max_eigen_vector = std::get<1>(*j);
+
+    Eigen::Quaternion<Real> q;
+    q.w() = max_eigen_vector(0);
+    q.x() = max_eigen_vector(1);
+    q.y() = max_eigen_vector(2);
+    q.z() = max_eigen_vector(3);
```

Comment:
and this should just be 

```
Eigen::Quaternion<Real> q(std::get<1>(*j));
```

(at least I think that should work. Otherwise try

```
Eigen::Quaternion<Real> q(std::get<1>(*j)(0), std::get<1>(*j)(1), std::get<1>(*j)(2), std::get<1>(*j)(3));
```